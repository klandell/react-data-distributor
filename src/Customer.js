// @flow
import * as React from 'react';
import { Consumer } from './SupplyChainContext';
import type { Data } from './flowTypes';

type Props = {
  render: (Data) => React.Node,
};

const Customer = ({ render }: Props) => (
  <Consumer>
    {(data: Data) => render(data)}
  </Consumer>
);

export default Customer;
