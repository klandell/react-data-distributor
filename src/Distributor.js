import React from 'react';
import { Provider } from './SupplyChainContext';
import memoize from './memoize';
import prepareData from './prepareData';

/*
selector: string,
data: object, // eslint-disable-line react/forbid-prop-types
formatters: objectOf(func),
children: oneOfType([node, arrayOf(node)]).isRequired,
*/

// create a memoized function to extract the relevant
// data from the supplied data
const getData = memoize(prepareData);

// a provider to pass data to the interested parties
// via the react context api
const Distributor = (props) => {
  const {
    selector,
    data,
    formatters,
    children,
  } = props;

  return (
    <Provider value={getData(selector, data, formatters)}>
      {children}
    </Provider>
  );
};

Distributor.defaultProps = {
  selector: '',
  data: {},
  formatters: {},
};

export default Distributor;
