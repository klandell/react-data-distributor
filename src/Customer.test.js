import React from 'react';
import { mount } from 'enzyme';
import Distributor from './Distributor';
import Customer from './Customer';

const data = {
  foo: 'foo',
  bar: {
    baz: 'baz',
  },
};

const formatters = {
  default: s => s.split('').join(' '),
  uppercase: s => s.toUpperCase(),
};

describe('<Customer>', () => {
  it('should render', () => {
    const wrapper = mount((
      <Distributor>
        <Customer render={() => <div />} />
      </Distributor>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive parcels', () => {
    const wrapper = mount((
      <Distributor data={data}>
        <Customer render={parcels => (
          <div>
            <span>{parcels.foo.value()}</span>
            <span>{parcels.bar.baz.value()}</span>
          </div>
        )} />
      </Distributor>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive formatters', () => {
    const wrapper = mount((
      <Distributor data={data} formatters={formatters}>
        <Customer render={parcels => (
          <div>
            <span>{parcels.foo.value('uppercase')}</span>
            <span>{parcels.bar.baz.value('uppercase')}</span>
          </div>
        )} />
      </Distributor>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive a default formatters', () => {
    const wrapper = mount((
      <Distributor data={data} formatters={formatters}>
        <Customer render={parcels => (
          <div>
            <span>{parcels.foo.value()}</span>
            <span>{parcels.bar.baz.value()}</span>
          </div>
        )} />
      </Distributor>
    ));
    expect(wrapper).toMatchSnapshot();
  });

  it('should receive a raw value', () => {
    const wrapper = mount((
      <Distributor data={data} formatters={formatters}>
        <Customer render={parcels => (
          <div>
            <span>{parcels.foo.rawValue}</span>
            <span>{parcels.bar.baz.rawValue}</span>
          </div>
        )} />
      </Distributor>
    ));
    expect(wrapper).toMatchSnapshot();
  });
});
