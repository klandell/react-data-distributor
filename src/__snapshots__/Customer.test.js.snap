// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Customer> should receive a default formatters 1`] = `
<Distributor
  data={
    Object {
      "bar": Object {
        "baz": "baz",
      },
      "foo": "foo",
    }
  }
  formatters={
    Object {
      "default": [Function],
      "uppercase": [Function],
    }
  }
>
  <Customer
    render={[Function]}
  >
    <div>
      <span>
        f o o
      </span>
      <span>
        b a z
      </span>
    </div>
  </Customer>
</Distributor>
`;

exports[`<Customer> should receive a raw value 1`] = `
<Distributor
  data={
    Object {
      "bar": Object {
        "baz": "baz",
      },
      "foo": "foo",
    }
  }
  formatters={
    Object {
      "default": [Function],
      "uppercase": [Function],
    }
  }
>
  <Customer
    render={[Function]}
  >
    <div>
      <span>
        foo
      </span>
      <span>
        baz
      </span>
    </div>
  </Customer>
</Distributor>
`;

exports[`<Customer> should receive formatters 1`] = `
<Distributor
  data={
    Object {
      "bar": Object {
        "baz": "baz",
      },
      "foo": "foo",
    }
  }
  formatters={
    Object {
      "default": [Function],
      "uppercase": [Function],
    }
  }
>
  <Customer
    render={[Function]}
  >
    <div>
      <span>
        FOO
      </span>
      <span>
        BAZ
      </span>
    </div>
  </Customer>
</Distributor>
`;

exports[`<Customer> should receive parcels 1`] = `
<Distributor
  data={
    Object {
      "bar": Object {
        "baz": "baz",
      },
      "foo": "foo",
    }
  }
  formatters={Object {}}
>
  <Customer
    render={[Function]}
  >
    <div>
      <span>
        foo
      </span>
      <span>
        baz
      </span>
    </div>
  </Customer>
</Distributor>
`;

exports[`<Customer> should render 1`] = `
<Distributor
  data={Object {}}
  formatters={Object {}}
>
  <Customer
    render={[Function]}
  >
    <div />
  </Customer>
</Distributor>
`;
